### PR TITLE
acceptance: get farmer to check state of load generators

### PR DESCRIPTION
**Pull request by vivekmenezes**
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/12230)

### CODE REVIEW

File: pkg/acceptance/terrafarm/farmer.go [hunk 1]
```diff
@@ -333,7 +338,15 @@ func (f *Farmer) Restart(ctx context.Context, i int) error {
 
 // Start starts the given process on the ith node.
 func (f *Farmer) Start(ctx context.Context, i int, process string) error {
-	_, _, err := f.execSupervisor(f.Nodes()[i], "start "+process)
+	name := "start " + process
+	_, _, err := f.execSupervisor(f.Nodes()[i], name)
+	if err == nil {
+		extendBy := f.NumNodes() - len(f.processes)
+		if extendBy > 0 {
```
Comment (tamird) [inline pkg/acceptance/terrafarm/farmer.go]: wouldn't all this new logic be simpler if you made `nodes` a `map[string][]string`?

Comment (vivekmenezes) [inline pkg/acceptance/terrafarm/farmer.go]: yes that's a good suggestion

**Comment by vivekmenezes** [comment]
This is ready for another review. I modified the nodes field to be a list of structs. The order of nodes matters or else we could have use a map too, but the struct is better over the long term where we could associate more state with a node.

**Comment by tamird** [comment]
Reviewed 1 of 2 files at r1.
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/acceptance/terrafarm/farmer.go, line 74 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12230#-KYfObQA6H0wQMzZdBDW:-KYfObQA6H0wQMzZdBDX:b-w0ba23) ([raw file](https://github.com/cockroachdb/cockroach/blob/51db7e5fe8089fe311be06c8e7dca4b7fc2a180a/pkg/acceptance/terrafarm/farmer.go#L74)):*
> ```Go
> 
> func (f *Farmer) refresh() {
> 	if len(f.nodes) != 0 {
> ```

this seems sketchy. perhaps you could make this really refresh things by running `supervisorctl status | grep -F RUNNING | cut -f1 -d' '` on each host.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12230)*

**Comment by cuongdo** [comment]
I must be 

---

Reviewed 1 of 2 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/acceptance/terrafarm/farmer.go, line 74 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12230#-KYfObQA6H0wQMzZdBDW:-KYnn0YDVTW2T0VkABIj:b1onloo) ([raw file](https://github.com/cockroachdb/cockroach/blob/51db7e5fe8089fe311be06c8e7dca4b7fc2a180a/pkg/acceptance/terrafarm/farmer.go#L74)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this seems sketchy. perhaps you could make this really refresh things by running `supervisorctl status | grep -F RUNNING | cut -f1 -d' '` on each host.
</blockquote></details>

Agreed, one expects to `refresh` to actually refresh something.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12230#-:-KYnnPUXzamMhSYIlyMt:b49yag2)*

**Comment by cuongdo** [comment]
(forget that last partial comment)

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12230#-:-KYnpBfC4nwPz3dL47Xf:bdwxrqi)*

**Comment by vivekmenezes** [comment]
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/acceptance/terrafarm/farmer.go, line 74 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12230#-KYfObQA6H0wQMzZdBDW:-KYoZoT_DoJZgMycDInZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/51db7e5fe8089fe311be06c8e7dca4b7fc2a180a/pkg/acceptance/terrafarm/farmer.go#L74)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary><blockquote>

Agreed, one expects to `refresh` to actually refresh something.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12230)*

**Comment by tamird** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/acceptance/terrafarm/farmer.go, line 77 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12230#-KYo_Lqf7L9lTf19goPf:-KYo_Lqf7L9lTf19goPg:b-5lk5b6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e5c1053d82f6c635666f65e6c714b546bcd40c2/pkg/acceptance/terrafarm/farmer.go#L77)):*
> ```Go
> 	f.nodes = make([]node, len(hosts))
> 	for i, host := range hosts {
> 		out, _, err := f.execSupervisor(host, "status | grep -F RUNNING | cut -f1 -d' '")
> ```

hope you've tested this!

---

*[pkg/acceptance/terrafarm/farmer.go, line 79 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12230#-KYo_b0KiBCg21grq_lt:-KYo_b0KiBCg21grq_lu:b-rdxxsr) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e5c1053d82f6c635666f65e6c714b546bcd40c2/pkg/acceptance/terrafarm/farmer.go#L79)):*
> ```Go
> 		out, _, err := f.execSupervisor(host, "status | grep -F RUNNING | cut -f1 -d' '")
> 		if err != nil {
> 			panic(err)
> ```

consider returning an error

---

*[pkg/acceptance/terrafarm/farmer.go, line 88 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12230#-KYobQKi3fTGjXIRkPnt:-KYobQKi3fTGjXIRkPnu:br4byvm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e5c1053d82f6c635666f65e6c714b546bcd40c2/pkg/acceptance/terrafarm/farmer.go#L88)):*
> ```Go
> }
> 
> func (f *Farmer) initNodes() {
> ```

consider making this method return a slice - that way all the methods below can chain.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12230#-:-KYobjJscIDUux9Ugk-Q:bnfp4nl)*

**Comment by cuongdo** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/acceptance/terrafarm/farmer.go, line 77 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12230#-KYo_Lqf7L9lTf19goPf:-KYobvba4kcIBpB0aIAz:bvsnld) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e5c1053d82f6c635666f65e6c714b546bcd40c2/pkg/acceptance/terrafarm/farmer.go#L77)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

hope you've tested this!
</blockquote></details>

+1 I'm mildly surprised this works. Combining pipes and SSH commands has usually caused me grief.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12230#-:-KYoc3z9kWzCeKAsET8S:bnfp4nl)*

**Comment by vivekmenezes** [comment]
Review status: 1 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/acceptance/terrafarm/farmer.go, line 77 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12230#-KYo_Lqf7L9lTf19goPf:-KYovPuy5NdBN6lldfqL:bht8zhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e5c1053d82f6c635666f65e6c714b546bcd40c2/pkg/acceptance/terrafarm/farmer.go#L77)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary><blockquote>

+1 I'm mildly surprised this works. Combining pipes and SSH commands has usually caused me grief.
</blockquote></details>

This works FWIW. I also tried a version where the post processing is done in GO, but the above version is short and sweet.

---

*[pkg/acceptance/terrafarm/farmer.go, line 79 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12230#-KYo_b0KiBCg21grq_lt:-KYoxxeOHE-u-OINUSt4:byfjuht) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e5c1053d82f6c635666f65e6c714b546bcd40c2/pkg/acceptance/terrafarm/farmer.go#L79)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

consider returning an error
</blockquote></details>

I've left this as is

---

*[pkg/acceptance/terrafarm/farmer.go, line 88 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12230#-KYobQKi3fTGjXIRkPnt:-KYoxxCFgYU8tm7VxBqM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e5c1053d82f6c635666f65e6c714b546bcd40c2/pkg/acceptance/terrafarm/farmer.go#L88)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

consider making this method return a slice - that way all the methods below can chain.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12230)*

**Comment by tamird** [comment]
lgtm, minor comments

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/acceptance/terrafarm/farmer.go, line 88 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12230#-KYobQKi3fTGjXIRkPnt:-KYp-K967gIRqouu5KjE:bh37fbn) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e5c1053d82f6c635666f65e6c714b546bcd40c2/pkg/acceptance/terrafarm/farmer.go#L88)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Done.
</blockquote></details>

perhaps this method should have a different name now.

---

*[pkg/acceptance/terrafarm/farmer.go, line 97 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12230#-KYp-QE0p_ktwhrcwTWb:-KYp-QE0p_ktwhrcwTWc:b-ydluwt) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbd6c876e7a9de227fe63e8217de53a7722f45d6/pkg/acceptance/terrafarm/farmer.go#L97)):*
> ```Go
> // FirstInstance returns the address of the first instance.
> func (f *Farmer) FirstInstance() string {
> 	return f.initNodes()[0].hostname
> ```

this is `return f.Hostname(0)`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12230#-:-KYp-UP3tDoFkWzMGU8k:b-k068r8)*

**Comment by vivekmenezes** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/acceptance/terrafarm/farmer.go, line 88 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12230#-KYobQKi3fTGjXIRkPnt:-KYpD9IhiW9M5NWmBdja:b-1lpz6c) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e5c1053d82f6c635666f65e6c714b546bcd40c2/pkg/acceptance/terrafarm/farmer.go#L88)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

perhaps this method should have a different name now.
</blockquote></details>

Actually Nodes should just be removed. I'll send you another PR with that.

---

*[pkg/acceptance/terrafarm/farmer.go, line 97 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12230#-KYp-QE0p_ktwhrcwTWb:-KYpD6kyJXfBHHiyiFt7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbd6c876e7a9de227fe63e8217de53a7722f45d6/pkg/acceptance/terrafarm/farmer.go#L97)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this is `return f.Hostname(0)`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12230)*

**Pull request merged**